### PR TITLE
Iox time and date functions

### PR DESCRIPTION
**Pull request by lwandzura**
Please let me know if I have missed any functions.  I know gapfill is not yet implemented.

### CODE REVIEW

File: content/influxdb/cloud-iox/reference/sql/functions/time-series.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+```sql
+SELECT date_part('hour', TIMESTAMP '2019-08-01T10:00:00Z') hour, 
+  "level description", 
+  "location",
+  time
+FROM "h2o_feet"
+WHERE time >= timestamp '2019-08-01T00:00:00Z' AND time <= timestamp '2019-10-31T00:00:00Z'
+ORDER BY time
+```
```
Comment (sanderson) [inline content/influxdb/cloud-iox/reference/sql/functions/time-series.md:212]:
~~~suggestion
```sql
SELECT
  date_part('hour', time) AS hour,
  time,
  "level description", 
  location
FROM h2o_feet
WHERE
  time >= timestamp '2019-08-17T02:54:00Z'
  AND time <= timestamp '2019-08-17T03:06:00Z'
ORDER BY time
```
~~~

File: content/influxdb/cloud-iox/reference/sql/functions/time-series.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+| hour | level description    | location     | time                     |
+| :--- | :------------------- | :----------- | :----------------------- |
+| 10   | below 3 feet         | santa_monica | 2019-08-17T00:00:00.000Z |
+| 10   | between 6 and 9 feet | coyote_creek | 2019-08-17T00:00:00.000Z |
+| 10   | below 3 feet         | santa_monica | 2019-08-17T00:06:00.000Z |
```
Comment (sanderson) [inline content/influxdb/cloud-iox/reference/sql/functions/time-series.md:218]:
```suggestion
| hour | time                 | level description    | location     |
| :--: | :------------------- | :------------------- | :----------- |
|  2   | 2019-08-17T02:54:00Z | between 3 and 6 feet | coyote_creek |
|  2   | 2019-08-17T02:54:00Z | between 3 and 6 feet | santa_monica |
|  3   | 2019-08-17T03:00:00Z | between 3 and 6 feet | coyote_creek |
|  3   | 2019-08-17T03:00:00Z | between 3 and 6 feet | santa_monica |
|  3   | 2019-08-17T03:06:00Z | between 3 and 6 feet | coyote_creek |
|  3   | 2019-08-17T03:06:00Z | between 3 and 6 feet | santa_monica |
```

**Review by sanderson — CHANGES_REQUESTED** [review]
A few notes:

- These functions aren't unique to time series, so I'm suggesting we call them "time and date" functions.
- Lets list the functions in alphabetical order
- Include a list of links at the top of the page that links to all the function headings
- There are more time and date functions listed in the DataFusion documentation that are missing here:
  - to_timestamp
  - to_timestamp_millis
  - to_timestamp_micros
  - to_timestamp_seconds
  - extract
  - from_unixtime

**Comment by alamb** [comment]
Please ping me again when @sanderson 's comments have been addressed. I don't want to end up reviewing content that is rewritten

**Comment by lwandzura** [comment]
I asked Andrew about some of the temporal functions I left out and while they do work, I am not sure what the use case would be for some, with the exception of extract.  But I can certainly include them.

**Comment by alamb** [comment]
> I asked Andrew about some of the temporal functions I left out and while they do work, I am not sure what the use case would be for some, with the exception of extract. But I can certainly include them.



I agree `extract ` has a usecase.

**Comment by lwandzura** [comment]
@alamb I am a little confused on the to_timestamp function usage

This works:

SELECT to_timestamp(time) 
FROM cpu 
LIMIT 1

However, the format in Datafusion's docs state this should work and I am getting errors:

basic syntax:  to_timestamp(timestamp, format)


SELECT TO_TIMESTAMP('2023-01-01 08:50:10', 'YYYY-MM-DD')
FROM h2o_feet

SELECT TO_TIMESTAMP('2020-01-31T09:26:56.123Z', 'YYYY-MM-DD')
FROM h2o_feet

Am I on the wrong track here?

**Comment by alamb** [comment]
Hi @lwandzura 

> However, the format in Datafusion's docs state this should work and I am getting errors:
> basic syntax: to_timestamp(timestamp, format)

The docs I can find don't refer to the second argument (though they are also not very helpful in general):
https://arrow.apache.org/datafusion/user-guide/expressions.html#temporal-expressions

I don't think the to_timestamp function supports a second argument currently in DataFusion -- can you copy the exact error you are seeing?

**Comment by alamb** [comment]
When I run this I see

```
❯ SELECT TO_TIMESTAMP('2020-01-31T09:26:56.123Z', 'YYYY-MM-DD');
Plan("Coercion from [Utf8, Utf8] to the signature Uniform(1, [Int64, Timestamp(Nanosecond, None), Timestamp(Microsecond, None), Timestamp(Millisecond, None), Timestamp(Second, None), Utf8]) failed.")
❯
```

Which says to me that the function requires a single argument of one of the type in that list (which is a pretty bad message)

**Review by sanderson — CHANGES_REQUESTED** [review]

**Pull request merged**